business/istio_validations.go

```diff
@@ -35,7 +35,7 @@ func (in *IstioValidationsService) GetServiceValidations(namespace, service stri
 
 	// Get Gateways and ServiceEntries to validate VirtualServices
 	wg := sync.WaitGroup{}
-	errChan := make(chan error, 2)
+	errChan := make(chan error, 64)
```

Comment:
I wonder if we need to refactor this method.
To me it starts to get complex and hard to follow.
I understand that increasing this to 64 it's because we don't control that max # of invocations could be 2 and we need some kind of higher level.
But it so, perhaps it's a matter to reorganize how are we grouping these methods or to use just a semaphore with a shared variable, I don't know, just thinking loud.

Reply:
I'll answer the previous question also. The question is that if we would have 65 namespaces and fetching data from each of them would fail at the same time, then it's possible that the last fetch is blocked indefinitely as the channel is never read before it is being written to.

In theory that could happen, but it would probably require large amount of bad luck with a scheduler + very beefy machine with large amount of cores.

Now, we could use something like the "context" that would cancel after first error and then prevent writing to it. Theoretically we could also do changes to the err-variable and let just any of them win (who cares which), but test -race does not like that although it's quite valid in this case. Maybe Golang has Java style atomic swap for variable also, I can't remember right now (at least unsafe has, but I don't remember if otherwise there is).

Also, sadly the goroutine can't return anything to the caller so err can't be handled that way (and channel can't be expanded on the fly). 

But all these methods are extremely complicated already and duplicating code, so a refactor is clearly necessary. Maybe it should be done in the context of KIALI-1671.

Reply:
> But all these methods are extremely complicated already and duplicating code, so a refactor is clearly necessary. Maybe it should be done in the context of KIALI-1671.

+1

Yes this logic has grown very fast and we need to revisit it.
KIALI-1671 sounds the right place to address it.